robotoff/spellcheck/patterns/__init__.py

```diff
@@ -0,0 +1,38 @@
+from typing import Dict
+from pathlib import Path
+
+from robotoff.spellcheck.base_spellchecker import BaseSpellchecker
+
+Patterns = Dict[str, str]
+
+
+class PatternsSpellchecker(BaseSpellchecker):
+
+    PATTERNS_PATHS = {"fr": Path(__file__).parent / "patterns_fr.txt"}
+
+    def __init__(self, lang: str = "fr"):
+        self.lang = lang
+        self.patterns: Patterns = {}
+
+        with self.PATTERNS_PATHS[lang].open() as f:
+            current_pattern = None
+            for line in f.readlines():
+                line = line.strip().split("#")[0]
+                if len(line) == 0:
+                    current_pattern = None
+                elif current_pattern is None:
+                    current_pattern = line
+                else:
+                    self.patterns[line.lower()] = current_pattern.lower()
+                    self.patterns[line.upper()] = current_pattern.upper()
+                    self.patterns[line.capitalize()] = current_pattern.capitalize()
+                    self.patterns[line] = current_pattern
+
+    @property
+    def name(self):
+        return super(PatternsSpellchecker, self).name + "__" + self.lang
+
+    def correct(self, text: str) -> str:
+        for key, value in self.patterns.items():
```

Comment:
I'm a bit concerned about the time complexity of this function, as it grows linearly with the number of patterns. You can either use a single regex (with `|` as OR operator) or flashtext whose complexity do not grow linearly with the number of patterns.

Reply:
I understand you concern. I'll see what I can do using a big regex instead of 1 `str.replace` per declared pattern. I really like flashtext but I don't think it's suitable in our case here. Flashtext looks for entire words in a sentence while I would like to look for patterns. For instance flashtext would miss the `ﬁ` character in a given word or `oeuf` in the sentence `blanc d'oeuf`. But I would totally sign for a flashtext working at a subword level.